grouper/fe/handlers/github.py

```diff
@@ -0,0 +1,144 @@
+import binascii
+import hmac
+import json
+import os
+from typing import cast, TYPE_CHECKING
+
+from six.moves.urllib.parse import urlencode
+from tornado import gen, httpclient
+
+from grouper.constants import USER_METADATA_GITHUB_USERNAME_KEY
+from grouper.fe.settings import settings
+from grouper.fe.util import GrouperHandler
+from grouper.models.audit_log import AuditLog
+from grouper.models.user import User
+from grouper.plugin import get_plugin_proxy
+from grouper.user_metadata import set_user_metadata
+
+if TYPE_CHECKING:
+    from grouper.models.base.session import Session
+    from tornado.concurrent import Future
+    from typing import Any, Iterator, Optional
+
+
+def _get_github_client():
+    # type: () -> httpclient.AsyncHTTPClient
+    return httpclient.AsyncHTTPClient()
+
+
+class GitHubLinkView(GrouperHandler):
+    @gen.coroutine
+    def get(self, *args, **kwargs):
+        # type: (*Any, **Any) -> Iterator[Future]
+        code = self.get_query_argument("code", None)
+        if code is None:
+            # Redirect the user to GitHub to authorize our app.
+            state = binascii.hexlify(os.urandom(16))
+            self.set_cookie("github-link-state", state, httponly=True)
+            params = {
+                "client_id": settings().github_app_client_id,
+                "state": state,
+                "redirect_uri": "{}/github/link/{}".format(settings().url, kwargs["user_id"]),
+            }
+            self.redirect("https://github.com/login/oauth/authorize?" + urlencode(params))
+            return
+
+        # Check that the state parameter is correct.
+        state = self.request.query_arguments.get("state", [b""])[-1]
+        expected_state = self.get_cookie("github-link-state", "").encode("utf-8")
+        self.clear_cookie("github-link-state")
+        if not hmac.compare_digest(state, expected_state):
+            self.badrequest()
+            return
+
+        # Make sure we're modifying the authenticated user before doing more.
+        user_id = kwargs["user_id"]
+        user = User.get(self.session, user_id)
+        if not user:
+            self.notfound()
+            return
+        if self.current_user.id != user.id:
+            self.forbidden()
+            return
+
+        # Turn the temporary oauth code into an access token.
+        github_client = _get_github_client()
```

Comment:
You're thinking something like this?
```
class GitHub:
    def login(self):...
    def get_username(self, token):...
```
I want the fake testing GitHub interface to be raw HTTP not something like above. I don't think Merou needs a generic GitHub client.

Normally, I would put what's currently in `get()` into a separate function and pass the HTTP client to it, so patching wasn't required. However, in this case, I want everything from the tornado web-frameworky code to the GitHub HTTP client integration tested.

In summary, this code was written in a deliberately inflexible way to minimize the places it could bend and break.

Reply:
Yep, that's roughly what I had in mind. I think we can keep the tests the same and still do this refactor? It would make the code more readable and maintainable.

Reply:
Sure, see latest commit.

Reply:
Awesome, thanks Benjamin!